File: scripts/bench_echoes/bench_layered_alv_step.jl
# =============================================================================
#  bench_layered_alv_step.jl — Julia counterpart of `bench_layered_alv_step.py`.
#  Tests step-activated layers (the pattern in script 37 :layers).
# =============================================================================

import Pkg
Pkg.activate(joinpath(@__DIR__, "..", ".."); io = devnull)

using MeanFieldHom
using TensND
using LinearAlgebra
using JSON3
using Printf

const HERE = @__DIR__
const json_path = joinpath(HERE, "bench_layered_alv_step_python.json")
isfile(json_path) || error("Run `bench_layered_alv_step.py` first")
ref = JSON3.read(read(json_path, String))

T_grid = collect(Float64.(ref.T))
n = length(T_grid)
const k0    = Float64(ref.k0)
const mu0   = Float64(ref.mu0)
const eta0  = Float64(ref.eta0)
const gamma0 = Float64(ref.gamma0)
const k1    = Float64(ref.k1)
const mu1   = Float64(ref.mu1)
const eta1  = Float64(ref.eta1)
const gamma1 = Float64(ref.gamma1)
const kp    = Float64(ref.kp)
const mup   = Float64(ref.mup)
const fp    = Float64(ref.fp)
const finf  = Float64(ref.finf)
const f0    = Float64(ref.f0)
const N     = Int(ref.N)
const t0    = Float64(ref.t0)
const lT    = collect(Float64.(ref.lT))

const C_p_tens = TensISO{3}(3 * kp, 2 * mup)

make_R0() = maxwell_iso(k0, mu0, eta0, gamma0)
make_R1() = maxwell_iso(k1, mu1, eta1, gamma1)

# Layer modulus matching Python visco_prop with `tp >= lt` history-dep:
function inclusion_law(t_set::Real)
    R1 = make_R1()
    return ViscoLaw(function (t, tp)
        if t < tp
            return zero(C_p_tens)
        elseif tp ≥ t_set
            return R1.eval_fun(t, tp)
        else
            return C_p_tens
        end
    end, :relaxation)
end

# Layers: pore + N solidifying shells.
# Outer-most carries lT[0] (innermost solidifying = lT[1] in Julia, earliest setting).
# Python: prop[0]=Cp, prop[i+1]=lT[N-1-i].  In Julia 1-indexed:
#   k=1 → pore; k=2..N+1 → lT[N-(k-2)] = lT[N - k + 2].
moduli = ntuple(N + 1) do k
    if k == 1
        heaviside_law(C_p_tens)
    else
        inclusion_law(lT[N - k + 2])
    end
end

cumulative = cumsum(vcat([fp], fill(finf / N, N)))
radii = ntuple(k -> cumulative[k]^(1 / 3), N + 1)

println("Setting times lT = ", lT)
println("Cumulative fractions: ", cumulative)
println("Radii: ", radii)

sphere = LayeredSphere(radii, moduli)
C0_law = make_R0()

α_jl = bulk_localization_alv(sphere, C0_law, T_grid)
β_jl = shear_localization_alv(sphere, C0_law, T_grid)

# Compare against Python reference per layer.
to_matrix(v) = reduce(hcat, [collect(Float64.(row)) for row in v])'

function compare(label, A_jl, A_py)
    @printf "\n── %s ──\n" label
    diff = A_jl .- A_py
    rel = maximum(abs, diff) / max(maximum(abs, A_py), 1e-300)
    @printf "  max |Δ| = %.3e   max relative = %.3e\n" maximum(abs, diff) rel
    println("  jl:")
    for i in 1:size(A_jl, 1)
        @printf "    [%d, :] = %s\n" i join([@sprintf("%+.5e", x) for x in A_jl[i, :]], "  ")
    end
    println("  py:")
    for i in 1:size(A_py, 1)
        @printf "    [%d, :] = %s\n" i join([@sprintf("%+.5e", x) for x in A_py[i, :]], "  ")
    end
end

for k in 1:(N + 1)
    α_py = Matrix(to_matrix(ref.layers_alpha[k]))
    β_py = Matrix(to_matrix(ref.layers_beta[k]))
    compare("Layer $k α", α_jl[k], α_py)
    compare("Layer $k β", β_jl[k], β_py)
end

File: scripts/bench_echoes/bench_layered_alv_step.py
"""
bench_layered_alv_step.py — cross-check layered-sphere ALV with the
step-activation pattern of `fluage_echoes_solid.py`.

Setup mirrors script 37 but with a small N (3 layers) and 5 time steps
so the (n×n) Volterra matrices are printable.  At t₀ = 0.5, layer 1
(innermost) is already activated (t_set ≈ 0.46), layer 2 not yet
(t_set ≈ 0.79).
"""

from numpy import *
from echoes import *
import json
set_printoptions(precision=8, suppress=True, linewidth=200)

# ── Material setup matching script 37 ─────────────────────────────────────

E0 = 1.0; nu0 = 0.2; C0 = stiff_Enu(E0, nu0)
f0 = 0.6; eta0 = 0.2; gamma0 = 0.133
E1 = 5.0; nu1 = 0.3; C1 = stiff_Enu(E1, nu1)
finf = 0.3; eta1 = 1.0; gamma1 = 1.67
Ep = E0 * 1.0e-8; nup = 0.2; Cp = stiff_Enu(Ep, nup); fp = 1 - f0 - finf

R0 = lambda t, tp: 3*C0.k*exp(-(t-tp)/eta0)*J4 + 2*C0.mu*exp(-(t-tp)/gamma0)*K4
R1 = lambda t, tp: 3*C1.k*exp(-(t-tp)/eta1)*J4 + 2*C1.mu*exp(-(t-tp)/gamma1)*K4

# Solidification kinetics
N = 3
alpha = 4.0
F = array([(i+0.5)*finf/N for i in range(N)])
lT = array([(f/(finf-f))**(1.0/alpha) for f in F])
print("setting times lT =", lT)

# Time grid (small).
T = array([0.5, 0.7, 1.0, 1.5, 2.0])
nT = len(T)
t0 = T[0]
print("t0 =", t0)

# Build layered RVE matching script 37 :layers.
ver = rve(matrix="MATRIX")
ver["MATRIX"] = ellipsoid(spherical, fraction=f0,
                          prop={"C": C0},
                          visco_prop={"C": (R0, RELAXATION)})

# Layers: pore + N solidifying shells.
# Outer-most is layer N (innermost solidifying = lT[0]); innermost is
# pore.  Lambda capture trick: lt=lT[N-1-i].
ver["INCLUSION"] = sphere_nlayers(
    radius=1.0,
    layer_fractions=[fp] + [finf/N for _ in range(N)],
    fraction=finf+fp,
    prop={"C": [Cp] + [C1 if t0>=lT[N-1-i] else Cp for i in range(N)]},
    visco_prop={"C": [(lambda _,__: Cp.array, RELAXATION)] +
        [(lambda t, tp, lt=lT[N-1-i]: R1(t, tp) if tp>=lt else Cp.array, RELAXATION)
         for i in range(N)]},
)

# Trigger MT.
V = homogenize_visco(prop="C", rve=ver, time_series=T, scheme=MT,
                     maxnb=100, epsrel=1.e-8, verbose=False)

print("\n=== Per-layer alpha (3K) and beta (2mu) Volterra n×n matrices ===")
out = {"T": T.tolist(), "lT": lT.tolist(), "t0": t0,
       "k0": C0.k, "mu0": C0.mu, "eta0": eta0, "gamma0": gamma0,
       "k1": C1.k, "mu1": C1.mu, "eta1": eta1, "gamma1": gamma1,
       "kp": Cp.k, "mup": Cp.mu, "fp": fp, "finf": finf, "f0": f0,
       "N": N, "alpha": alpha,
       "layers_alpha": [], "layers_beta": []}
for k in range(N+1):
    raw = ver["INCLUSION"].layer_visco_eE(k)
    a, b = visco_paramsym(raw, ISO)
    print(f"\n--- Layer {k} ---")
    print("alpha:")
    print(a)
    print("beta:")
    print(b)
    out["layers_alpha"].append(a.tolist())
    out["layers_beta"].append(b.tolist())

# Effective creep compliance.
J = linalg.inv(V)
JE = sum(J[::6, ::6], 1)
print("\nEffective creep compliance J^E_eff(t, t0):")
print(JE)
out["JE_eff"] = JE.tolist()

import os
out_path = os.path.join(os.path.dirname(__file__), "bench_layered_alv_step_python.json")
with open(out_path, "w") as f:
    json.dump(out, f, indent=2)
print(f"\nSaved Python step reference to {out_path}")
